### PR TITLE
Fix fused MLA and MTP checkpointing for Megatron-FSDP.

### PR DESCRIPTION
**Pull request by rapatel**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Fixes fused MLA and MTP checkpointing for Megatron-FSDP.

Aside from the newly added unit tests, this PR was also tested manually for save/load parity check for the fused MLA + MTP path:

* Ran a checkpoint save/load round trip on 2 GPUs through the real code in this PR (preprocess_fsdp_dtensor_state_dict -> the MLA/MTP handlers -> split_fused_fsdp_param), with a real torch.distributed.checkpoint save and load and the new validate_fsdp_dtensor_model_load. All checks passed.
* The model reproduces: a fused linear_qkv_down_proj with asymmetric sections, an input_layernorm absorbed as layer_norm_weight, and an MTP layer whose inner layer is mtp_model_layer in the model but transformer_layer on disk, and which itself contains a fused attention, so the MLA-before-MTP handler ordering is covered.

The 3 main checks were:
1. Unfused checkpoint -> fused model
    - 16 of 16 tensors bit-identical to what was saved, 0 mismatches across ranks
    - loss 1.2154645919799805 vs 1.2154645919799805, delta 0.000e+00, bit-identical
2. Fused -> fused round trip (destination model starts from a different seed)
    - 16 of 16 on-disk tensors bit-identical
    - 13 of 13 weights the forward pass reads bit-identical
    - loss 1.1673169136047363 before save and after load, bit-identical
    - all gradients bit-identical, e.g. linear_qkv_down_proj.weight gradient norm 0.16416169703006744 before and after
3. Negative control, identical load with the MLA handler disabled
    - CheckpointingException raised, naming the 6 unloadable fused parameters
    - with --dist-ckpt-strictness ignore_all, the same load silently keeps those 6 at their initial values, which is the behavior that broke convergence before this change

* Additionally, we ran a 64-node (256 GB300 GPUs) DSv3 convergence run with these changes, which  ran as expected within the target samples-to-converge range for the target GBS.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/config/training_config.py [hunk 1]
```diff
@@ -571,6 +571,24 @@ class CheckpointConfig:
...
+    ] = "raise_unexpected"
+    """Determine handling of model weights an FSDP DTensor partial load cannot supply.
+    Check StrictHandling docs for flags meaning. Only applies when strict_fsdp_dtensor_load is
+    False, where the underlying load would otherwise skip such weights silently and leave them
+    at their initialized values. Unlike strict_fsdp_dtensor_load this covers only the model
+    sections, so optimizer and RNG state that is intentionally absent does not trip it. Weights
+    the checkpoint holds but this rank does not need are always fine, as expected under
+    pipeline, tensor and expert parallelism. Ignored when finetune is set."""
+
     dist_ckpt_strictness: Literal[
```
Comment (cspades) [inline megatron/training/config/training_config.py]: Can we just pipe the existing `dist_ckpt_strictness` argument to FSDP?

File: megatron/core/transformer/fsdp_dtensor_checkpoint.py [hunk 2]
```diff
@@ -564,6 +566,315 @@ def split_gdn_fused(data, dist_param, split_sizes, split_dim):
...
+    while parts and parts[0] in ('module', 'model'):
+        parts = parts[1:]
+    return '.'.join(parts)
+
+
+def _intersect_slice(s1, s2):
+    """Intersection of two step-1 slices, or an empty slice when they do not overlap."""
+    start = max(s1.start, s2.start)
+    stop = min(s1.stop, s2.stop)
+    return slice(0, 0) if start >= stop else slice(start, stop)
```
Comment (cspades) [inline megatron/core/transformer/fsdp_dtensor_checkpoint.py]: There's a similar function in the SwiGLU transforms, I think this can be used for that.

**Review by cspades — COMMENTED** [review]
I think it LGTM (there's a lot of code but a good bit is familiar code). Have 1 comment on arguments, and also an ask that you check we at least have loss and grad parity loading a non-trivial checkpoint with the model this PR is supporting, thanks!

File: megatron/training/config/training_config.py (hunk 1 above)
Comment (rapatel) [reply megatron/training/config/training_config.py]: Done.

File: megatron/core/transformer/fsdp_dtensor_checkpoint.py (hunk 2 above)
Comment (rapatel) [reply megatron/core/transformer/fsdp_dtensor_checkpoint.py]: Done.

**Comment by rapatel** [comment]
@cspades Please see updated description of this PR - it includes the details from the save/load parity testing. Thanks.

**Review by cspades — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]

**Review by asolergi-nv — APPROVED** [review]

**Comment by rapatel** [comment]
@claude I've addressed your concerns. But this needs CI to run the functional tests to generate the gold values. I've added two checkpoint-resume cases covering both fixed paths, registered in tests/test_utils/recipes/gb200/moe-1node.yaml

**Review by asolergi-nv — APPROVED** [review]

**Review by jiemingz — APPROVED** [review]

**Review by maanug-nv — APPROVED** [review]
